src/components/navBar/navBar.js

```diff
@@ -103,23 +106,68 @@ function MdNavBar($mdAria) {
       'mdSelectedNavItem': '=?',
       'navBarAriaLabel': '@?',
     },
-    template:
-      '<div class="md-nav-bar">' +
-        '<nav role="navigation">' +
-          '<ul class="_md-nav-bar-list" layout="row" ng-transclude role="listbox"' +
-            'tabindex="0"' +
-            'ng-focus="ctrl.onFocus()"' +
-            'ng-blur="ctrl.onBlur()"' +
-            'ng-keydown="ctrl.onKeydown($event)"' +
-            'aria-label="{{ctrl.navBarAriaLabel}}">' +
-          '</ul>' +
-        '</nav>' +
-        '<md-nav-ink-bar></md-nav-ink-bar>' +
-      '</div>',
-    link: function(scope, element, attrs, ctrl) {
+    template: '<div class=\'md-nav-wrapper\'> ' +
+                '<div class=\'md-nav-pagination\'> ' +
+                  '<md-prev-button ' +
+                  'aria-label="Previous Page" ' +
+                  'aria-disabled="{{!ctrl.canPageBack()}}" ' +
+                  'ng-disabled="{{!ctrl.canPageBack()}}" ' +
+                  'ng-if="ctrl.shouldPaginate" ' +
+                  'ng-click="ctrl.previousPage()"> ' +
+                    '<md-icon md-svg-src="' + $$mdSvgRegistry.mdTabsArrow + '"></md-icon> ' +
+                  '</md-prev-button> ' +
+                  '<md-next-button ' +
+                  'aria-label="Next Page" ' +
+                  'aria-disabled="{{!ctrl.canPageForward()}}" ' +
+                  'ng-disabled="{{!ctrl.canPageForward()}}" ' +
+                  'ng-if="ctrl.shouldPaginate" ' +
+                  'ng-click="ctrl.nextPage()"> ' +
+                    '<md-icon md-svg-src="' + $$mdSvgRegistry.mdTabsArrow + '"></md-icon> ' +
+                  '</md-next-button> ' +
+                '</div>' +
+                '<div class="md-nav-bar" ' +
+                'md-swipe-left="ctrl.nextPage()" ' +
+                'md-swipe-right="ctrl.previousPage()"> ' +
+                  '<nav role="navigation"> ' +
+                    '<ul class="_md-nav-bar-list" ng-transclude ' +
+                    'layout="row" ' +
+                    'role="listbox"' +
+                    'tabindex="0"' +
+                    'ng-focus="ctrl.onFocus()" ' +
+                    'ng-blur="ctrl.onBlur()" ' +
+                    'ng-keydown="ctrl.onKeydown($event)" ' +
+                    'aria-label="{{ctrl.navBarAriaLabel}}"> ' +
+                    '</ul> ' +
+                  '</nav> ' +
+                  '<md-nav-ink-bar></md-nav-ink-bar> ' +
+                '</div> ' +
+              '</div> ',
+    link: function (scope, element, attrs, ctrl) {
       if (!ctrl.navBarAriaLabel) {
         $mdAria.expectAsync(element, 'aria-label', angular.noop);
       }
+
+      var observer = new MutationObserver(function (mutations) {
+        ctrl.updatePagination();
```

Comment:
I'm not sure there's a **good** way to solve it, since we still need some calls to `clientWidth`. Trying to avoid the `canPaginate` calls as much as possible, as well as caching the results` should help.